core/trace/src/main/scala/org/typelevel/otel4s/trace/TraceState.scala

```diff
@@ -0,0 +1,238 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.trace
+
+import cats.Show
+import cats.kernel.Hash
+
+/** An '''immutable''' representation of the key-value pairs defined by the W3C
+  * Trace Context specification.
+  *
+  * Trace state allows different vendors propagate additional information and
+  * interoperate with their legacy Id formats.
+  *
+  *   - Key is opaque string up to 256 characters printable. It MUST begin with
+  *     a lowercase letter, and can only contain lowercase letters a-z, digits
+  *     0-9, underscores _, dashes -, asterisks *, and forward slashes /.
+  *
+  *   - Value is opaque string up to 256 characters printable ASCII RFC0020
+  *     characters (i.e., the range 0x20 to 0x7E) except comma , and =.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/trace/api/#tracestate]]
+  *   [[https://www.w3.org/TR/trace-context/#mutating-the-tracestate-field]]
+  *   [[https://www.w3.org/TR/trace-context/#tracestate-header-field-values]]
+  */
+sealed trait TraceState {
+
+  /** Returns the value to which the specified key is mapped, or `None` if this
+    * map contains no mapping for the key.
+    *
+    * @param key
+    *   with which the specified value is to be associated
+    */
+  def get(key: String): Option[String]
+
+  /** Adds or updates the entry that has the given `key` if it is present. If
+    * either the key or the value is invalid, the entry will not be added.
+    *
+    * @param key
+    *   the key for the entry to be added. Key is an opaque string up to 256
+    *   characters printable. It MUST begin with a lowercase letter, and can
+    *   only contain lowercase letters a-z, digits 0-9, underscores _, dashes -,
+    *   asterisks *, and forward slashes /. For multi-tenant vendor scenarios,
+    *   an at sign (@) can be used to prefix the vendor name. The tenant id
+    *   (before the '@') is limited to 240 characters and the vendor id is
+    *   limited to 13 characters. If in the multi-tenant vendor format, then the
+    *   first character may additionally be numeric.
+    *
+    * @param value
+    *   the value for the entry to be added. Value is opaque string up to 256
+    *   characters printable ASCII RFC0020 characters (i.e., the range 0x20 to
+    *   0x7E) except comma , and =.
+    *
+    * @return
+    *   a new instance of [[TraceState]] with added entry
+    */
+  def put(key: String, value: String): TraceState
```

Comment:
I wonder if there are better names we can use for `put` and `remove` to reflect that it is immutable e.g. how Scala `Map` uses `updated`. But I don't have a strong opinion.

Reply:
What if we use `removed` and `updated` instead? As you mentioned, it should be closer to Scala Map.